ext/dom/html_document.c

```diff
@@ -0,0 +1,1047 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Niels Dossche <nielsdos@php.net>                            |
+   +----------------------------------------------------------------------+
+*/
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "php.h"
+#if defined(HAVE_LIBXML) && defined(HAVE_DOM)
+#include "php_dom.h"
+#include "html5_parser.h"
+#include "html5_serializer.h"
+#include "namespace_compat.h"
+#include <Zend/zend_smart_string.h>
+#include <lexbor/html/encoding.h>
+#include <lexbor/encoding/encoding.h>
+
+/* Implementation defined, but as HTML5 defaults in all other cases to UTF-8, we'll do the same. */
+#define DOM_FALLBACK_ENCODING_NAME "UTF-8"
+#define DOM_FALLBACK_ENCODING_ID LXB_ENCODING_UTF_8
+
+typedef struct {
+	size_t last_line;
+	size_t last_column;
+	size_t last_offset;
+} dom_line_column_cache;
+
+typedef struct {
+	const char *input_name;
+	const lxb_codepoint_t *current_input_codepoints;
+	const char *current_input_characters;
+	size_t current_input_length;
+	size_t current_total_offset;
+	dom_line_column_cache cache_tokenizer;
+	bool html_no_implied;
+} dom_lexbor_libxml2_bridge_application_data;
+
+typedef struct {
+	const lxb_encoding_data_t *encoding_data;
+	size_t bom_shift;
+} dom_character_encoding_data;
+
+typedef zend_result (*dom_write_output)(void*, const char *, size_t);
+
+typedef struct {
+	const lxb_encoding_data_t *encoding_data;
+	const lxb_encoding_data_t *decoding_data;
+	lxb_encoding_encode_t *encode;
+	lxb_encoding_decode_t *decode;
+	lxb_codepoint_t *codepoints;
+	lxb_char_t *encoding_output;
+	void *output_data;
+	dom_write_output write_output;
+} dom_output_ctx;
+
+typedef struct {
+	/* We can skip some conversion if the input and output encoding are both UTF-8, we only have to validate and substitute replacement characters */
+	bool fast_path; /* Put first, close to the encode & decode structures, for cache locality */
```

Comment:
```suggestion
	bool fast_path; /* Put first, near the encode & decode structures, for cache locality */
```

Okay, this is a dumb nit, but I initial read this as the verb "closing" which confused me... thanks English.